engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/DefinitionArgument.scala

```diff
@@ -216,6 +216,36 @@ object DefinitionArgument {
     override def children: List[IR] =
       name :: ascribedType.toList ++ defaultValue.toList
 
+    override def withNewChildren(newChildren: List[IR]): IR = {
```

Comment:
I have to admit that I was horrified by the body of this method when I saw it for the first time. Probably this is the reason why I want to avoid `withNewChildren` for now and I am proposing to [rely just on IR.mapExpressions for now](https://github.com/enso-org/enso/pull/11191/files#r1787612365).

#### Why is this method so complicated?

Because we loose all type information and we aren't sure what constructor to call, right? If we only traverse via  [IR.mapExpressions](https://github.com/enso-org/enso/pull/11191/files#r1787612365). Then this kind of problem is avoided.

In all places where `ir.Expression` is used, it can easily be replaced by any other subclass of `ir.Expression`. This is not the case for `IR` in general. If there is `ir.Name` field then one cannot replace it for `ir.Expression` or anything else than `ir.Name`.

The way we do the transformation right now - e.g. via `IR MiniIRPass.transformIr(IR)` we cannot enforce type safety of such a transformations. It may happen one gets in `IR.Name` are returns `IR` of unsuitable type. Thus, for now I would just change to `ir.Expression transformIr(ir.Expression)` and use `IR.mapExpressions` without introducing any `withNewChildren`.

Reply:
At the end e2fc538 keeps signature of the `transformIr` class and just throws runtime exception when expecting `Expression` to in `mapExpressions` transformation.

The good thing on `IR transformIr(IR)` is that it works on `IR.Module` and can thus modify imports, exports and bindings.

Reply:
Right now the PR doubles the `prepare` and `transform` methods as @Akirathan preferred that. I don't, but it is probably not that essential. Just `MiniIRPass.compile` has to [do the switch](https://github.com/enso-org/enso/pull/11191/commits/8124e1e1e6738f1829f1554d8456604bb1d26d3a#diff-8cb76fd855b3002b301048af801552074db5c9016bc28d0e261fdc044840b338R47) instead of leaving that up to the actual implementation.